### PR TITLE
rebase--interactive: rewrite complete_action() in C

### PR DESCRIPTION
**Pull request by agrn**
This rewrites complete_action() from shell to C. The relevant commits are the last two.

### CODE REVIEW

**Comment by agrn** [comment]
@dscho @stefanbeller @chriscool

File: git-rebase--interactive.sh [hunk 1]
```diff
@@ -127,54 +127,6 @@ init_revisions_and_shortrevisions () {
...
-
-$comment_char $(eval_ngettext \
-	"Rebase \$shortrevisions onto \$shortonto (\$todocount command)" \
-	"Rebase \$shortrevisions onto \$shortonto (\$todocount commands)" \
-	"$todocount")
-EOF
-	git rebase--helper --append-todo-help ${keep_empty:+--keep-empty}
-
-	has_action "$todo" ||
-		return 2
```
Comment (chriscool) [inline git-rebase--interactive.sh]: Why isn't there code for the above 2 lines in the function you wrote?

File: git-rebase--interactive.sh [hunk 2]
```diff
@@ -127,54 +127,6 @@ init_revisions_and_shortrevisions () {
...
-
-	cp "$todo" "$todo".backup
-	collapse_todo_ids
-	git_sequence_editor "$todo" ||
-		die_abort "$(gettext "Could not execute editor")"
-
-	has_action "$todo" ||
-		return 2
```
Comment (chriscool) [inline git-rebase--interactive.sh]: Here also why isn't there code for the above 2 lines in the function you wrote?

File: sequencer.c [hunk 3]
```diff
@@ -4374,11 +4377,9 @@ int check_todo_list(void)
 	return res;
 }
 
-int append_todo_help(unsigned edit_todo, unsigned keep_empty)
+static void append_todo_help(unsigned edit_todo, unsigned keep_empty,
+			     struct strbuf *buf)
```
Comment (chriscool) [inline sequencer.c]: I think it would be easier to review if changes to this function were in a separate commit.

**Comment by admin7770** [comment]
Danke ;)

Christian Couder <notifications@github.com> schrieb am Fr., 22. Juni 2018,
09:26:

File: sequencer.c [hunk 4]
```diff
@@ -4479,7 +4473,7 @@ static int rewrite_file(const char *path, const char *buf, size_t len)
 }
 
 /* skip picking commits whose parents are unchanged */
-int skip_unnecessary_picks(void)
+static int skip_unnecessary_picks(const char **output_oid)
```
Comment (chriscool) [inline sequencer.c]: Maybe changes to this function could also go to another commit.

**Comment by admin7770** [comment]
Ok sorry

Christian Couder <notifications@github.com> schrieb am Fr., 22. Juni 2018,
09:32:

File: builtin/rebase--helper.c [hunk 5]
```diff
@@ -41,21 +43,17 @@ int cmd_rebase__helper(int argc, const char **argv, const char *prefix)
 			N_("expand commit ids in the todo list"), EXPAND_OIDS),
 		OPT_CMDMODE(0, "check-todo-list", &command,
 			N_("check the todo list"), CHECK_TODO_LIST),
-		OPT_CMDMODE(0, "skip-unnecessary-picks", &command,
-			N_("skip unnecessary picks"), SKIP_UNNECESSARY_PICKS),
```
Comment (chriscool) [inline builtin/rebase--helper.c]: Maybe removel of useless modes could be done in another commit.

**Comment by chriscool** [comment]
There is "rebase--helper: remove checkout-onto" after the Signed-off-by at the end of the commit that rewrites complete_action().

File: git-rebase--interactive.sh (hunk 1 above)
Comment (agrn) [reply git-rebase--interactive.sh]: Return code 2 is a special case for interactive rebase. It tells git-rebase.sh to apply the autostash, to remove the state dir and to die with the message “nothing to do” ([see here](https://github.com/git/git/blob/master/git-rebase.sh#L215-L220)). I decided to rewrite this case in C, so instead of `return 2`, there is [this](https://github.com/git/git/pull/513/files/d77c4414835ec158c87d863e849f1c5f9c322a6d..0b4b1c0fb3cd9bbc80f1914ebbebf2a81b38886d#diff-79231f0693f84f3951daeea17065aad9R4615).

After the call to `--append-todo-help`, I noticed that the todo-list has at least one action: if the todo file is empty, then complete_action() writes a noop there. So I removed this specific check from my code.

Comment (agrn) [reply git-rebase--interactive.sh]: It also reminds me that I forgot to remove the handling of return code 2 in git-rebase.sh.

Comment (agrn) [reply git-rebase--interactive.sh]: Oh no, it would break git-rebase--preserve-merges.sh.

File: git-rebase--interactive.sh (hunk 2 above)
Comment (agrn) [reply git-rebase--interactive.sh]: Same as above, except here it is actually handled by my code [here](https://github.com/git/git/pull/513/files/d77c4414835ec158c87d863e849f1c5f9c322a6d..0b4b1c0fb3cd9bbc80f1914ebbebf2a81b38886d#diff-79231f0693f84f3951daeea17065aad9R4613).

**Comment by admin7770** [comment]
Thank you :)

agrn <notifications@github.com> schrieb am Fr., 22. Juni 2018, 11:36:

File: git-rebase--interactive.sh (hunk 1 above)
Comment (chriscool) [reply git-rebase--interactive.sh]: Ok, please explain all that in the commit message.

**Comment by admin7770** [comment]
Ok Work Tal you

Christian Couder <notifications@github.com> schrieb am Fr., 22. Juni 2018,
11:45:

**Comment by agrn** [comment]
Updated. The relevant commits are the last four.

**Comment by chriscool** [comment]
Yeah, this looks better to me now.

In the commit message I would suggest to start with the more general information and to have details toward the end.

**Comment by agrn** [comment]
This should be better, then

**Comment by chriscool** [comment]
Yeah, this looks good to me! Thanks!

**Comment by dscho** [comment]
@agrn this PR should be addressed by b49ef560ed66, yes?
